DI registrations

jeremydmiller opened the pull request:
The new `AddMarten()` service registrations we've been talking about for v3.12. I have not finished the documentation yet, but the code is ready to look at.

I'm punting on the async daemon to v4 if that's okay. 

I'll be adding more docs in my tomorrow. Sorry, but that was just a bit too much tonight.

oskardudycz commented on src/Marten.Testing/MartenServiceCollectionExtensionsTests.cs:
```diff
@@ -0,0 +1,142 @@
+
+namespace Marten.Testing
+{
+
```
**SUGGESTION:** It would be nice to remove those empty lines.

jeremydmiller replied:
Let's please not invest time in this kind of code formatting in the test code

oskardudycz commented on src/Marten/MartenServiceCollectionExtensions.cs:
```diff
@@ -0,0 +1,144 @@
+            services.AddSingleton<IDocumentStore, DocumentStore>();
+
+            // This can be overridden by the expression following
+            services.AddSingleton<ISessionFactory, DefaultSessionFactory>();
```
**SUGGESTION:** We should use `TryAddSingleton` method to allow user to override this implemntation. See more in [docs](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.dependencyinjection.extensions.servicecollectiondescriptorextensions.tryaddsingleton?view=dotnet-plat-ext-3.1).
```suggestion
            services.TryAddSingleton<ISessionFactory, DefaultSessionFactory>();
```

jeremydmiller replied:
I don't think that's necessary here Oskar.

oskardudycz commented on src/Marten/MartenServiceCollectionExtensions.cs:
```diff
@@ -0,0 +1,144 @@
+            /// <returns></returns>
+            public MartenConfigurationExpression BuildSessionsWith<T>() where T : class, ISessionFactory
+            {
+                _services.AddSingleton<ISessionFactory, T>();
```
Same comment as above https://github.com/JasperFx/marten/pull/1501/files#r436521253.

Also `ISessionFactory` not always should be Singleton. For a non-default case I'd rather expect it to be Scoped - eg. if the user would like to start the session with Tenant. The tenant probably will be taken from the request.

jeremydmiller commented:
Hey all, I think I've addressed all the comments in the PR and I want to declare victory. I've:

* flushed out the guide for the new extensions
* added a super, quick sample to the Getting Started page and linked to the bigger guide on AddMarten()
* replaced the StructureMap example with Lamar as the alternative IoC registration

mysticmind commented:
This is super good docs, thanks @jeremydmiller ❤️ 

Also this addresses the most frequent question asked by users!
